### PR TITLE
Framework: Move to Node 4

### PR DESCRIPTION
**Pull request by blowery**
This moves us up to Node v4.2.3, the current LTS version, in the Docker container. This needs a fair bit of testing and should not be merged until we're pretty happy with it.

Yes, the branch name is wrong.

To test, you'll need to 
1. install Vagrant, 
2. add a `env-config.sh` to the root of the project which exports CALYPSO_ENV set to a environment you want to test
3. run `vagrant up`
4. add the correct entries to /etc/hosts to hit your local env for the host you want to pretend to be
5. add in a forwarder from port 80 to port 3000 locally, using nginx or polipo

TODO:
- [x] Bump Dockerfile up to 4.2.3
- [x] Add `circle.yml` to tell Circle CI which node to use
- [x] Upgrade any packages that need updates to run well on 4
- [x] Fix up tests
- [x] Bump `engines` in `package.json`
- [x] Add `Makefile` target that checks node version and explodes if the dev is on pre 4.2.3.

### CODE REVIEW

**Comment by ehg** [comment]
It would be cool to merge https://github.com/Automattic/wp-calypso/pull/730 to avoid some duplication. 

I will give 4.4.2 a test with vagrant. With regards to merging this, I think a similar approach to the `webpack` server bundler is a good idea. e.g. notify systems, disable deploys to production for 30 minutes.

**Comment by blowery** [comment]
One little wrinkle I ran into: on Node v4+, it appears that the default http listener only listens on an IPv6 loopback, if it's available. If not, it falls back to the normal 127.0.0.1 IPv4 address. I had to update my /etc/hosts mapping to point at `::1` instead of `127.0.0.1` to make things work as expected.

**Comment by blowery** [comment]
Bah, I transposed the node version on the branch name and the commit. Should be 4.2.2, not 4.4.2.

Fixing the commit.

**Comment by blowery** [comment]
Redid this on top of master now that #730 landed

**Comment by blowery** [comment]
Hmmmm... So switching Circle CI over to node 4.2.2 breaks the build, even if built with a clean cache. It appears to be breaking on contextify? Maybe jumping up to a current jsdom would help there. It's pretty disconcerting to see all the build errors.

**Comment by ehg** [comment]
> Maybe jumping up to a current jsdom would help there. It's pretty disconcerting to see all the build errors.

@sirbrillig I think you mentioned the latest jsdom doesn't need `contextify`?

**Comment by sirbrillig** [comment]
Yup, that's correct! [Relevant comment](https://github.com/tmpvar/jsdom/blob/69dbd06a695912c13fc8ee519b36cfdb153a344c/Changelog.md#300).

> 3.0.x will be the last release of jsdom to support Node.js. All future releases (starting with 4.0.0) will require io.js, whose new vm module will allow us to remove our contextify native-module dependency. (Given that I submitted the relevant patch to joyent/node 1.5 years ago, I'm very excited that we can finally use it!)

**Comment by ehg** [comment]
So we'd have to get everyone to upgrade to `node` 4+ on their local machines… :)

File: client/my-sites/upgrades/cart/test/test-cart-buttons.jsx [hunk 1]
```diff
@@ -3,7 +3,7 @@ require( 'lib/react-test-env-setup' )();
 /**
  * External Dependencies
  */
-var	sinon = require( 'sinon' ),
+const	sinon = require( 'sinon' ),
```
Comment (aduth) [inline client/my-sites/upgrades/cart/test/test-cart-buttons.jsx]: Does this change need to be made? If we're changing things, I'd rather see this go to `import` syntax than to `const`.

Comment (blowery) [inline client/my-sites/upgrades/cart/test/test-cart-buttons.jsx]: it was mostly to quiet up eslint.

**Comment by gwwar** [comment]
Don't forget to update the node engine version to our minimum supported version in `package.json`:
```
"engines": {
    "node": ">=4.2.3"
}
```

**Comment by blowery** [comment]
> Don't forget to update the node engine version to our minimum supported version in package.json

Done. Interesting reading at http://www.marcusoft.net/2015/03/packagejson-and-engines-and-enginestrict.html. Looks like it's mostly advisory since we're the app.

**Comment by gwwar** [comment]
Yup, it sadly only enforces it if you install as a dependency, but as a node developer it's still the first place I check for version compatibility.

**Comment by sirbrillig** [comment]
A version check in the Makefile might make sense, as suggested elsewhere by @jkudish. Something like this as a target:

```make
node-version:
	@if [ "$(shell $(NODE) --version | sed 's/[^0-9]//g')" -lt 400 ]; then echo "Please upgrade your version of Node.js: https://nodejs.org/"; exit 1; fi
```

**Comment by blowery** [comment]
> A version check for Node in the Makefile might make sense,

I added this a slightly different way, using a node script and the semver property in package. This way it could work on Windows too, and we have a single source of truth for the necessary node version.

**Comment by sirbrillig** [comment]
> using a node script and the semver property in package

That's way better! Awesome. :+1:

**Comment by blowery** [comment]
Except semver isn't baked in... sigh.

**Comment by aduth** [comment]
I force-pushed a rebase to account for tests added in #550 which would have broken after the upgrade here (fixup'd into b81e2de).

I also fixed the version compare script to pull the `semver` dependency from the npm global root `node_modules` directory. The `semver` package is not bundled with Node, but is declared as a dependency of and maintained by the folks behind npm (9f8d852). This feels a bit hacky, but works well in my testing.

**Comment by blowery** [comment]
> This feels a bit hacky, but works well in my testing.

Another option would be to pull in semver to the checking script (or to ./bin) and use it directly. Not sure which is grosser. Having it on hand is less likely to break in the future when npm changes things.

**Comment by aduth** [comment]
>Not sure which is grosser.

I don't hold a strong opinion one way or the other, but with the approach in 9f8d852, I'm comforted by the fact that:

- We know that `npm` will always be an installed global package
- The `semver` package is [maintained by npm](https://docs.npmjs.com/misc/semver) and I'd find it very unlikely that it'd ever be dropped as a dependency from the `npm` package

Another approach that we discussed privately is to change the Makefile target to install `semver` prior to running the script:

```make
node-version:
    @$(NPM) install semver && $(BIN)/check-node-version
```

**Comment by blowery** [comment]
> Another approach that we discussed privately is to change the Makefile target to install semver prior to running the script:

Right and I think we both thought that was too slow to be reasonable, since it gets run as a dependency of install, which is a dep of everything. I'd rather not incur network traffic just to validate the version of node. 

Maybe we're being too fancy and we should just validate that the current version's major is at least 4. We can always change it down the road when we require 6. I'm sure someone will yell that it got out of date, and it's just a dev helper.

**Comment by sirbrillig** [comment]
As an aside, since the script that checks for the node version requires node to be installed, it might also be worth adding a Makefile target to be sure node is installed at all. I know it's in the README, but it would be one more nice thing for new devs.

One possible example:

```make
npm:
	@echo "Checking for npm..."
	@command -v $(NPM) >/dev/null 2>&1 || { echo >&2 "Please install Node.js: https://nodejs.org/"; exit 1;  }
```

**Comment by blowery** [comment]
Should we add a project-wide `.nvmrc` ? Would that be useful?

**Comment by aduth** [comment]
>Should we add a project-wide `.nvmrc` ? Would that be useful?

Sounds useful, though it would appear that we'd need to update Make targets to use `nvm exec` if `nvm` is installed? Personally, I use `n`, so I wouldn't benefit from this.

**Comment by blowery** [comment]
> Sounds useful, though it would appear that we'd need to update Make targets to use nvm exec if nvm is installed?

Nah, it just makes `nvm use` easier. No reason to force `nvm exec` into the Makefile.

**Comment by blowery** [comment]
Hrm, serving `/` and the CSS seems much slower now. I'm seeing the CSS rebuild on every request... Digging.

**Comment by blowery** [comment]
> Hrm, serving / and the CSS seems much slower now. I'm seeing the CSS rebuild on every request... 

This was due to the new `node-version` target. Made it order-only in 338e3e9 which prevents it from causing a rebuild on every request.

**Comment by blowery** [comment]
Just got this up and running, pretending to be a production node, and it seems to work just fine. I think this is ready to go.

**Comment by ehg** [comment]
I think it's good to go too, works fine for me :+1:

**Pull request merged**